Make destroy async transactional

gwincr11 opened the pull request:
Motivation:
  - A active record rollback could occur while enqueuing a job in this
    case the job would enqueue even though the database deletion
    rolledback putting things in a funky state. So this adds some
    functionality to ensure that deletions are not queued until after
    the db transaction has been commited. Ideally a 
    `has_many destroy: :async` should handle this for the user.

Related Issues:
  - Sort of related to https://github.com/rails/rails/issues/26045

Changes:
  - add some syntatic sugar to transactions to make enqueueing a job a
    little better experience, this really hooks into the after destroy
    hook on the transaction.
  - Setup async destroy on associations to not enqueue until after a
    transaction has been committed.

gwincr11 committed:
Make destroy async transactional
Motivation:
  - A active record rollback could occur while enqueuing a job in this
    case the job would enqueue even though the database deletion
    rolledback putting things in a funky state. So this adds some
    functionality to ensure that deletions are not queued until after
    the db transaction has been commited.

Related Issues:
  - Sort of related to https://github.com/rails/rails/issues/26045

Changes:
  - add some syntatic sugar to transactions to make enqueueing a job a
    little better experience, this really hooks into the after destroy
    hook on the transaction.
  - Setup async destroy on associations to not enqueue until after a
    transaction has been committed.

MatheusRich commented on activerecord/lib/active_record/transactions.rb:
```diff
@@ -209,6 +209,10 @@ def transaction(**options, &block)
         connection.transaction(**options, &block)
       end
 
+      def transactional_job_enqueue(*args, &block)
```
Will this warn in Ruby 2.7?

rafaelfranca commented on activerecord/lib/active_record/transactions.rb:
```diff
@@ -209,6 +209,10 @@ def transaction(**options, &block)
         connection.transaction(**options, &block)
       end
 
+      def transactional_job_enqueue(*args, &block)
```
Why is this method necessary? I see no reason to a new method to the public API to do this.

gwincr11 replied:
I am just sort of spit balling this, I was not really sure where this should be or how the api should work so I put it here to get conversation going.

rafaelfranca commented on activerecord/lib/active_record/associations/association.rb:
```diff
@@ -331,7 +331,10 @@ def skip_statement_cache?(scope)
         def enqueue_destroy_association(options)
-          owner.class.destroy_association_async_job&.perform_later(**options)
+          job_class = owner.class
+          ActiveRecord::Base.transactional_job_enqueue options do
```
Would not this add an after commit callback to all the models in the application even if they don't have destroy async associations?

rafaelfranca reviewed (changes requested):
I don't think the implementation is correct. We want to start the job in an after commit, but only for models that have `destroy: :async` associations. They way it is implemented now it add a callback, at runtime, to all the models.

I think the callback should be registered only once, and the should add the job to a list that is then called in the callback.

Something like this that I extracted from Shopify (but not necessarially exactly like this):

```ruby
# typed: false
# frozen_string_literal: true
module QueueForAfterCommit
  extend ActiveSupport::Concern

  included do |base|
    base.after_commit(:execute_queued_for_later)
  end

  def run_after_commit(*args, &block)
    queued_for_later.push(*args)
    queued_for_later.push(block) if block
  end

  private

  def queued_for_later
    @queued_for_later ||= []
  end

  def execute_queued_for_later
    queued_for_later.uniq.each do |hook|
      if hook.respond_to?(:call)
        hook.call
      else
        send(hook)
      end
    end

    @queued_for_later = []
  end
end
```

gwincr11 committed:
add new module to handle queueing of jobs to enqueue after commit is successful

gwincr11 commented:
Thanks @rafaelfranca I took a shot at adapting that I am not sure how generic or defined that api should be please add thoughts. I am not sure if it will be public or private.

rafaelfranca committed:
Refactoring the code to not add methods to all models

Only models that need to use the `:destroy_async` feature
need those methods so it is better to not add to all models.
